### PR TITLE
bug fix: stop all kinds of expressions from cnf-exploding

### LINKED ISSUE #14569: Bug Report: Query rewrite leads to exponential growth in query text length
### Overview of the Issue

When `vtgate` cannot determine any VIndex to use for query routing in a sharded keyspace, it falls back to performing various sets of query rewrites that semantically don't make a difference in the query execution, but can help in determining a better query plan to use.

One of the rewrites it performs is trying to transform a query into [CNF](https://en.wikipedia.org/wiki/Conjunctive_normal_form). Unfortunately, that transform can lead to an explosion in query length, which is exponentially correlated to the number of `OR` expression the query planner tries to rewrite.

Even a low number `OR` expressions can easily cause a query to become hundred thousands of characters long, and we queries in production that contain 50 or more of these `OR` expressions, which make this rewrite extremely problematic.

Here's an example query:

```sql
SELECT COUNT(*)
FROM music
WHERE (music.user_id = 1
       AND music.name = 'test'
       OR music.user_id = 2
       AND music.name = 'test2'
       OR music.user_id = 3
       AND music.name = 'test3'
       OR music.user_id = 4
       AND music.name = 'test4'
       OR music.user_id = 5
       AND music.name = 'test5')
  AND music.other = FALSE
GROUP BY user_id,
         name
```

Here's the rewritten query text:

```sql
SELECT count(*)
FROM music
WHERE ((music.user_id = 1
        AND music.`name` = 'test'
        OR music.user_id = 2
        AND music.`name` = 'test2'
        OR music.user_id = 3)
       AND (music.user_id = 1
            AND music.`name` = 'test'
            OR music.user_id = 2
            AND music.`name` = 'test2'
            OR music.`name` = 'test3')
       OR music.user_id = 4
       AND music.`name` = 'test4'
       OR music.user_id = 5)
  AND ((music.user_id = 1
        AND music.`name` = 'test'
        OR music.user_id = 2
        AND music.`name` = 'test2'
        OR music.user_id = 3)
       AND (music.user_id = 1
            AND music.`name` = 'test'
            OR music.user_id = 2
            AND music.`name` = 'test2'
            OR music.`name` = 'test3')
       OR music.user_id = 4
       AND music.`name` = 'test4'
       OR music.`name` = 'test5')
  AND music.other = FALSE
GROUP BY user_id,
         `name`
```

### Reproduction Steps

N/A

### Binary Version

```sh
v16 and later
```

### Operating System and Environment details

```sh
N/A
```

### Log Fragments

```sh
N/A
```

### PR DESCRIPTION
**Pull request by systay**
## Description
In a recent change (#14560) we fixed a bug in the CNF rewriter that stopped it from rewriting a lot of cases.
Unfortunately, we succeeded too well, and the newly rewritten expressions are timing out in expression fuzzing on CI. This is because CNF rewriting in some situations result in huge outputs (see [Wikipedia](https://en.wikipedia.org/wiki/Conjunctive_normal_form#Conversion_into_CNF))

This PR stops the rewriting once the produced expression has grown to 10 times the size of the original expression. 

## Related Issue(s)
Fixes #14569

## Checklist
-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

**Review by vmg — CHANGES_REQUESTED** [review]
I think we need to look at benchmarks for this @systay, the O complexity of this new code is pathological because after each rewrite, you need to walk the whole AST to count the nodes.

Since we control the rewrite patterns, I think it would be much more efficient to keep track of the amount of added & removed nodes during the rewrite. The tracking doesn't need to be fully accurate, just enough to detect unbounded growth, so it should be very easy to implement!

**Comment by systay** [comment]
> I think we need to look at benchmarks for this @systay, the O complexity of this new code is pathological 

Yeah, you are right. This was sloppy. I'll work on it

File: go/vt/sqlparser/predicate_rewriting.go [hunk 1]
```diff
@@ -16,32 +16,21 @@ limitations under the License.
...
-		}
-
-		return true, nil
-	}, ast)
-
-	allowCNF := count < CNFOrLimit
+	original, initialSize := cloneAndSize(ast)
 
+	r := &rewriter{
+		initialSize: initialSize,
```
Comment (vmg) [inline go/vt/sqlparser/predicate_rewriting.go:25]: I don't think this field is being used.

File: go/vt/sqlparser/predicate_rewriting_test.go [hunk 2]
```diff
@@ -91,8 +91,14 @@ func TestSimplifyExpression(in *testing.T) {
 			expr, err := ParseExpr(tc.in)
 			require.NoError(t, err)
 
-			expr, didRewrite := simplifyExpression(expr, true)
-			assert.True(t, didRewrite.changed())
+			r := &rewriter{
+				// setting these number so it never stops rewriting for these tests
+				initialSize: 1000000000,
```
Comment (vmg) [inline go/vt/sqlparser/predicate_rewriting_test.go:96]: `math.MaxInt`? :smile:

**Review by vmg — APPROVED** [review]
aaaah linear time

image

File: go/vt/sqlparser/predicate_rewriting.go [hunk 3]
```diff
@@ -16,32 +16,14 @@ limitations under the License.
...
-
-		return true, nil
-	}, ast)
-
-	allowCNF := count < CNFOrLimit
+	original := CloneSQLNode(ast)
 
-	for {
-		printExpr(ast)
+	// This loop can in some cases of this conversion to CNF can lead to an exponential explosion of the formula,
```
Comment (arthurschreiber) [inline go/vt/sqlparser/predicate_rewriting.go:24]: Something's off with this comment 😬

**Review by dbussink — APPROVED** [review]

**Pull request merged**